src/generic_requirement_type.cpp

```diff
@@ -0,0 +1,297 @@
+}
+
+void generic_requirement_type::finalize()
+{
+
+}
+
+void generic_requirement_type::check() const
+{
+    for( const weather_type_id &weather : required_weathers ) {
+        if( !weather.is_valid() ) {
+            debugmsg( "weather type %s does not exist.", weather.c_str() );
+            abort();
+        }
+    }
+    for( const trait_id &trait : required_traits ) {
+        if( !trait.is_valid() ) {
+            debugmsg( "trait %s does not exist.", trait.c_str() );
+            abort();
+        }
+    }
+    for( const efftype_id &effect : required_effects ) {
+        if( !effect.is_valid() ) {
+            debugmsg( "effect %s does not exist.", effect.c_str() );
+            abort();
+        }
+    }
+    for( const bionic_id &cbm : required_cbms ) {
+        if( !cbm.is_valid() ) {
+            debugmsg( "cbm %s does not exist.", cbm.c_str() );
+            abort();
+        }
+    }
+}
+
+void generic_requirement_type::load( const JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "pressure_min", pressure_min, INT_MIN );
+    optional( jo, was_loaded, "pressure_max", pressure_max, INT_MAX );
+    optional( jo, was_loaded, "humidity_min", humidity_min, INT_MIN );
+    optional( jo, was_loaded, "humidity_max", humidity_max, INT_MAX );
+    optional( jo, was_loaded, "temperature_min", temperature_min, INT_MIN );
+    optional( jo, was_loaded, "temperature_max", temperature_max, INT_MAX );
+    optional( jo, was_loaded, "windpower_min", windpower_min, INT_MIN );
+    optional( jo, was_loaded, "windpower_max", windpower_max, INT_MAX );
+    optional( jo, was_loaded, "humidity_and_pressure", humidity_and_pressure,
+              true );
+    optional( jo, was_loaded, "time", time,
+              time_requirement_type::both );
+    for( const std::string &required_weather :
+         jo.get_string_array( "required_weathers" ) ) {
+        required_weathers.push_back( weather_type_id( required_weather ) );
+    }
+    for( const std::string &required_trait :
+         jo.get_string_array( "required_traits" ) ) {
+        required_traits.push_back( trait_id( required_trait ) );
+    }
+    for( const std::string &required_effect :
+         jo.get_string_array( "required_effects" ) ) {
+        required_effects.push_back( efftype_id( required_effect ) );
+    }
+    for( const std::string &required_cbm :
+         jo.get_string_array( "required_cbms" ) ) {
+        required_cbms.push_back( bionic_id( required_cbm ) );
+    }
+    for( const std::string &variable : jo.get_string_array( "true_generic_variables" ) ) {
+        true_generic_variables.push_back( variable );
+    }
+    for( const std::string &variable : jo.get_string_array( "false_generic_variables" ) ) {
+        false_generic_variables.push_back( variable );
+    }
+    optional( jo, was_loaded, "time_passed_min", time_passed_min,
+              0_seconds );
+    optional( jo, was_loaded, "time_passed_max", time_passed_max,
+              0_seconds );
```

Comment:
This should probably use `cata_optional` rather than using a sentinel value of `0_seconds`.